### PR TITLE
Code review recommended practices

### PR DESCRIPTION
**Pull request by aminespinoza**
# Pull Request Template

## What are you trying to address
This PR includes a document for code review best practices as recommendations to be implemented into projects when dealing with Pull requests and Code reviews in order to make teams more productive and faster when reviewing code.

No issues related to this PR.

## Description of new changes

- There is a link added in Code Reviews main documents to be able to see the new markdown file.
- There is a new file presenting the best practices gathered from internal CSE experiences and also from external sources about how to be more efficient in the code reviews.

### CODE REVIEW

**Review by katyamust — DISMISSED** [review]
This needs more work before it is merged. Lets sit down together and review it.

File: docs/code-reviews/best-practices.md
Comment (katyamust) [reply docs/code-reviews/best-practices.md:32]: Amin, @aminespinoza, were you able to look to comments and address them? If you feel this is too much changes for one PR we can try to split the changes into smaller ones together.  Let me know if I can help here.

**Comment by aminespinoza** [comment]
Updates made:
- Best practices original document deleted and content merged in Author Guidance and Process Guidance. 
- Broken link removed from pull-requests.md
- Reference link to original document deleted in code reviews/REDAME.md
- Mermaid diagram added to pull-requests.md

**Comment by aminespinoza** [comment]
- Modified link referring to NPM download in CONTRIBUTING.md

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/author-guidance.md:31]:
```suggestion
If you can't avoid large PRs, include explanations of the changes in order to make it easier for the reviewer to review the code, with clear comments the reviewer can identify the goal of every code block.
```

File: docs/code-reviews/process-guidance/README.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/README.md:33]: This sounds like an author specific guidance, can we move it there?

File: docs/code-reviews/process-guidance/README.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/README.md:31]:
```suggestion
Metrics can play a significant role in this part, you can include an **Inspection rate** (the speed with which review is performed), **Defect rate** (the number of bugs found within an hour of review) and **Defect density** (the number of bugs found per piece of code) as the most common ones to be considered.
```

File: docs/code-reviews/process-guidance/README.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/README.md:39]:
```suggestion
You shouldn't review code too quickly but neither take too long in one sitting. If you have many pull requests or if the complexity of code is demanding the recommendation is to take a break between reviews to recover, and focus one the ones you are most experienced with.
```

File: docs/code-reviews/process-guidance/README.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/README.md:37]: This sounds like a reviewer specific guidance, can we move it to the reviewer section?

File: docs/code-reviews/process-guidance/README.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/README.md:41]:
```suggestion
Always remember, the goal of a code review is to verify that the specified task has been achieved. If your suggestions are on code related to, but not included in the PR, raise those suggestions as separate tasks (bugs, technical debt etc.). Don't block the current PR on suggestions outside the scope of the PR.
```

File: docs/code-reviews/process-guidance/README.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/README.md:45]:
```suggestion
Before doing any code review it's important to set automated builds, tests and checks (something achievable in the [CI process](../../continuous-integration/README.md)), this can save human reviewers some time and focus in areas like design and functionality for proper evaluation.
```

Comment (TessFerrandez) [reply docs/code-reviews/process-guidance/README.md:45]: The last part was not related to the topic "automate whenever possible" and is also covered in the reviewer guidance already

File: docs/code-reviews/process-guidance/README.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/README.md:47]:
```suggestion
This will ensure the team is focusing on the right things and it will reduce review time.
```

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/author-guidance.md:27]:
```suggestion
At the moment of creating a pull request it is a good idea to put in the description of it a checklist of expected objectives to achieve in the PR. This way reviewers can be able to focus on specific areas and also consider some other tasks as secondary according to this list.
```

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (TessFerrandez) [reply docs/code-reviews/process-guidance/author-guidance.md:31]: Must say that I don't necessarily agree with this. The code should be able to speak for itself, through proper naming and through splitting large functions into smaller etc. If we need to add comments, the code should be refactored to be more readable. Comments will always get stale and incorrect over time as the code changes, but the devs forget about updating the comments.

We should however, add comments to the PR to describe the changes we have made, and why.

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/author-guidance.md:39]:
```suggestion
Code reviews play a critical role in product quality and it should not represent an arena for long discussions or even worse a battle of egos. What matters is a bug caught not who made it, not who found it, not who solved it. The only thing that matters is having the best possible product.
```

File: docs/code-reviews/pull-requests.md
Comment (TessFerrandez) [inline docs/code-reviews/pull-requests.md:88]:
```suggestion
* [Facebook approach to hidden features](https://launchdarkly.com/blog/secret-to-facebooks-hacker-engineering-culture/)
* [Azure approach to canary releases](https://docs.microsoft.com/en-us/azure/architecture/framework/devops/deployment#stage-your-workloads)
```

Comment (TessFerrandez) [reply docs/code-reviews/pull-requests.md:88]: I have fixed this in #775 - so that it points to the new correct location -- once this PR is ready for merge, let's merge in that change instead of removing this content

**Review by TessFerrandez — CHANGES_REQUESTED** [review]
Thank you for the changes, I suggested some edits to re-structure and avoid duplication
+ moving back the link so that we don't loose the content (the link is fixed in #775)

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (aminespinoza) [reply docs/code-reviews/process-guidance/author-guidance.md:31]: I like the improvement, change made!

File: docs/code-reviews/process-guidance/README.md
Comment (aminespinoza) [reply docs/code-reviews/process-guidance/README.md:33]: I took the section to the author guidance.

File: docs/code-reviews/process-guidance/README.md
Comment (aminespinoza) [reply docs/code-reviews/process-guidance/README.md:45]: Thanks I saw it already.

**Review by TessFerrandez — APPROVED** [review]

File: docs/code-reviews/process-guidance/README.md
Comment (TessFerrandez) [reply docs/code-reviews/process-guidance/README.md:27]: This was just resolved without action... was there a reason?

File: docs/code-reviews/process-guidance/reviewer-guidance.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/reviewer-guidance.md:37]: I think this fits better in the author guidance than in the reviewer as the author has the power over how big the PRs are

File: docs/code-reviews/process-guidance/README.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/README.md:33]: There is already a section on automation in the reviewer guidance (under "Decide on a common standard for each language"). Rather than duplicating this - can we consolidate the information in these sections? I personally think that the consolidated section fits better here than under the reviewer section (as it is general preparation the team needs to do - but we don't want to have duplicate sections)

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/author-guidance.md:43]: This is more a reviewer guidance than an author guidance - was it put here by mistake?

**Review by TessFerrandez — CHANGES_REQUESTED** [review]
Thank you for the changes, I think some sections may have ended up in the wrong files by mistake after the modifications -- please have a look through the comments below.

File: docs/code-reviews/process-guidance/README.md
Comment (aminespinoza) [reply docs/code-reviews/process-guidance/README.md:27]: It's already there.

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (aminespinoza) [reply docs/code-reviews/process-guidance/author-guidance.md:43]: Well, goal here was to let the author know it's ok to ping the reviewer and let him know there's some code there to be reviewed and probably unblock this in a faster way.

Comment (TessFerrandez) [reply docs/code-reviews/process-guidance/author-guidance.md:43]: Are we referring to the same section here? This was in regards to 

"Foster a positive code review culture

Code reviews play a critical role in product quality and it should not represent an arena for long discussions or even worse a battle of egos. What matters is a bug caught not who made it, not who found it, not who solved it. The only thing that matters is having the best possible product."

Comment (aminespinoza) [reply docs/code-reviews/process-guidance/author-guidance.md:43]: Well, not that you say it that way yes, it seems to fit.

File: docs/code-reviews/process-guidance/reviewer-guidance.md
Comment (aminespinoza) [reply docs/code-reviews/process-guidance/reviewer-guidance.md:37]: In fact, I think the entire section "Make comments clear" can fit in the author side. Gonna make that change and let me know what you think.

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/author-guidance.md:39]: This is reviewer guidance - not author guidance

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/author-guidance.md:31]: This should probably go in the general section

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (TessFerrandez) [inline docs/code-reviews/process-guidance/author-guidance.md:46]: This should probably go in the reviewer section - or in the general section

File: docs/code-reviews/process-guidance/reviewer-guidance.md
Comment (TessFerrandez) [reply docs/code-reviews/process-guidance/reviewer-guidance.md:30]: Looks like this was still not moved

**Review by TessFerrandez — CHANGES_REQUESTED** [review]

**Review by TessFerrandez — CHANGES_REQUESTED** [review]

**Review by tompaana — CHANGES_REQUESTED** [review]
After the suggestions by @TessFerrandez have been addressed, I can help with copy-editing to improve the language.

File: docs/code-reviews/process-guidance/author-guidance.md
Comment (aminespinoza) [reply docs/code-reviews/process-guidance/author-guidance.md]: After talking to some of our teammates we think it fits in author, we just gave it more context.

File: docs/code-reviews/pull-requests.md
Comment (aminespinoza) [reply docs/code-reviews/pull-requests.md:30]: Ok, changes added.

**Review by tompaana — APPROVED** [review]
LGTM!

**Review by ikivanc — APPROVED** [review]
Thanks Amin, it looks good to me.

**Review by TessFerrandez — APPROVED** [review]

**Pull request merged**
